packages/x-date-pickers/src/DateCalendar/useCalendarState.tsx

```diff
@@ -158,6 +160,17 @@ export const useCalendarState = <TDate extends PickerValidDate>(
     slideDirection: 'left',
   });
 
+  // Ensure that `calendarState.currentMonth` is updated when `referenceDate` (or timezone changes)
+  // https://github.com/mui/mui-x/issues/10804
+  React.useEffect(() => {
+    dispatch({
+      type: 'changeMonth',
+      newMonth: utils.startOfMonth(referenceDate),
```

Comment:
Are we sure this change is wanted?
From what I can understand it will reset the displayed month to `referenceDate`, so if the user had moved to another month and then the timezone changes it might be weird.

Tbh this is really an edge case and I could understand that we don't want to optimize for it


____

More problematic, this will break if the component re-renders and `referenceDate` is passed directly is not memoized (which we support for now and we should keep supporting IMHO):

```tsx
import * as React from 'react';
import dayjs from 'dayjs'
import Stack from "@mui/material/Stack";
import Switch from "@mui/material/Switch";
import { DemoContainer } from '@mui/x-date-pickers/internals/demo';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { DateCalendar } from '@mui/x-date-pickers/DateCalendar';

export default function BasicDateCalendar() {
    const [isActive, setIsActive] = React.useState(false)

    return (
        <Stack spacing={2}>
            <Switch value={isActive} onChange={event => setIsActive(event.target.checked)} />
            <LocalizationProvider dateAdapter={AdapterDayjs}>
                <DemoContainer components={['DateCalendar']}>
                    <DateCalendar referenceDate={dayjs('2022-04-17')} />
                </DemoContainer>
            </LocalizationProvider>
        </Stack>
    );
}
```

Reply:
@flaviendelangle WDYT about the updated solution: https://github.com/mui/mui-x/pull/12321/commits/f859853266c25cd9dd280ec84d968c6099f0387d?
It would update only the `timezone` and optionally in some cases the `month` (to avoid changing the month when switching timezone to `-the current one`). I.e. `America/New_York` -> `America/Los_Angeles`

Reply:
From what I understand, of the two problems I describe above, the 1st one still occurs but the 2nd does not, right?

Reply:
Were you able to replicate the first mentioned problem?
It does not happen to me. 🤔 
This part: https://github.com/mui/mui-x/pull/12321/files#diff-b671464bcd4d0027334e53087da5e27c98342ac1c23b1550a4556879d6f20300R64-R66 is to ensure that the month doesn't shift when `timezone` changes to a minus offset.

Reply:
[Example on master](https://stackblitz.com/edit/react-vsw4h7?file=Demo.tsx)

Change the month to be July using the arrows, then click on "Change timezone".
It should move back to April.

The funny thing is, we already have this behavior on master (I would have thought otherwise), so no regression in your PR :+1:

Reply:
That's a bit strange behavior that might be worth investigating a bit more in-depth.
I'm a bit puzzled why the current month doesn't change when the current month is changed to `May` (+1). 🤔

Reply:
Not a clue :laughing: 
If you can create an issue and add it to the board so that we don't loose track of it
I don't think it's the top priority right now but I agree it would be interesting to understand what is going on